### DISCUSSION TITLE
Cache size increases with each subsequent run

### DISCUSSION THREAD

**Discussion by mtsokol** [ccache/ccache#1529] [Support]
Hi!

I recently started using `ccache` to decrease build times of a custom MLIR dialect in GitHub Actions CI jobs. The build process requires building LLVM and a dialect. For LLVM build I use `ccache` for the compilation caching, with:
```
-DLLVM_CCACHE_BUILD=ON
or
-DCMAKE_C_COMPILER_LAUNCHER=ccache \
-DCMAKE_CXX_COMPILER_LAUNCHER=ccache
```

I noticed that with each subsequent run the cache size increases by ~30%, so my current approach requires to delete all caches in GitHub Action cache list from time to time to avoid exceeding the storage limit.

I can reproduce it locally and I noticed that after subsequent runs `.ccache/` directory has a few more cache files. Is it possible to tell `ccache` to remove unused cache files after a run? (I'd expect that for identical runs cache size stays the same).

I would like to explore this topic a bit more. Is it possible to inspect what is being cached additionally between runs?

Here's a `ccache` configuration that I'm using: https://github.com/finch-tensor/Finch-mlir/blob/412117b441dd1aad8b05f875267129ff57d0f17b/.github/workflows/ci.yml#L65
It's based on a `mlir-wheels` repository: https://github.com/makslevental/mlir-wheels/blob/main/actions/setup_ccache/action.yml
And here's a Actions' cache: https://github.com/finch-tensor/Finch-mlir/actions/caches

![image](https://github.com/user-attachments/assets/c2b84824-b3f6-4a07-b536-18c181360f19)

**Comment by jrosdahl** [comment]
> Is it possible to tell ccache to remove unused cache files after a run?

You can prune based on size or time. Size-based pruning happens automatically if you configure a max cache size. Time-based pruning can be done by calling e.g. `ccache --evict-older-than 3600s`.

> Is it possible to inspect what is being cached additionally between runs?

You can enable a [log file](https://ccache.dev/manual/4.10.2.html#config_log_file) or [stats log](https://ccache.dev/manual/4.10.2.html#config_stats_log) and look for `direct_cache_miss` and `preprocessed_cache_miss`. This will indicate the additional files.
